### PR TITLE
Update flutter/http.dart to use dart:io ... redux.

### LINKED ISSUE #5668: http response bodies are sometimes truncated
Run the stocks example 10 times on the iOS simulator and you'll likely get a failure along the lines of:

```
ERROR:../../lib/tonic/logging/dart_error.
[ERROR:../../lib/tonic/logging/dart_error.cc(16)] Unhandled exception:
FormatException: Unterminated string (at line 466, character 11)
    "$146.
          ^
#0      _ChunkedJsonParser.fail (dart:convert-patch/convert_patch.dart:1328)
#1      _ChunkedJsonParser.close (dart:convert-patch/convert_patch.dart:477)
#2      _parseJson (dart:convert-patch/convert_patch.dart:20)
#3      JsonDecoder.convert (dart:convert/json.dart:501)
#4      StockDataFetcher._fetchNextChunk.<anonymous closure> (/Users/drewwarren/flutter/examples/stocks/lib/stock_data.dart:82:38)
#5      _RootZone.runUnary (dart:async/zone.dart:1404)
#6      _FutureListener.handleValue (dart:async/future_impl.dart:131)
#7      _Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:637)
#8      _Future._propagateToListeners (dart:async/future_impl.dart:667)
```

I get the failure about 40% of the time. I have only tested this on the iOS simulator.

```
✓] Flutter (on Mac OS, channel master)
    • Flutter at /Users/drewwarren/flutter
    • Framework revision e0f6c0db84 (26 hours ago), 2016-08-29 12:17:34
    • Engine revision c4022b61fa
    • Tools Dart version 1.19.0-dev.5.0

[x] Android toolchain - develop for Android devices
    x Android Studio / Android SDK not found. Download from https://developer.android.com/sdk/
      (or visit https://flutter.io/setup/#android-setup for detailed instructions).

[✓] iOS toolchain - develop for iOS devices (Xcode 7.3.1)
    • XCode at /Applications/Xcode.app/Contents/Developer
    • Xcode 7.3.1, Build version 7D1014

[✓] Atom - a lightweight development environment for Flutter
    • flutter plugin version 0.2.4
    • dartlang plugin version 0.6.37
```

### LINKED ISSUE #5808: Deprecate & remove the Mojo based network stack.
See discussion in https://github.com/flutter/engine/pull/3008#discussion_r78233924 for details. Boils down to the following points:
* Users have an alternative in the form of `dart:io`.
* Implementing the network service correctly on each platform is difficult to do if the interfaces don't allow us to control how data is read from the underlying socket.
* If we cant control how we read from the socket, we have to buffer data. For large requests, we have to manage OOM conditions or crash.

### PR DESCRIPTION
**Pull request by chinmaygarde**
- Updates Flutter's copy of http.dart to match tip-of-tree with the use of mirrors removed.
- Removes unused NetworkAssetBundle and MojoAssetBundle.fromNetwork.
- We can now deprecate the Mojo network stack.
- Fixes https://github.com/flutter/flutter/issues/5808
- Fixes https://github.com/flutter/flutter/issues/5668
- Replaces https://github.com/flutter/engine/pull/3008

cc @abarth @jason-simmons @Hixie

### CODE REVIEW

**Comment by chinmaygarde** [comment]
Github wont let me re-open https://github.com/flutter/flutter/pull/5851 because I seemed to have force pushed to that branch at some point (even though I made no actual changes in the push).

File: packages/flutter/pubspec.yaml [hunk 1]
```diff
@@ -5,7 +5,9 @@ description: A framework for writing Flutter applications
 homepage: http://flutter.io
 
 dependencies:
+  async: "^1.10.0"
   collection: '>=1.9.1 <2.0.0'
+  http_parser: ">=0.0.1 <4.0.0"
```
Comment (abarth) [inline packages/flutter/pubspec.yaml]: Can you add a comment that we can remove these once we switch to using `package:http`?

Comment (chinmaygarde) [inline packages/flutter/pubspec.yaml]: Done

**Comment by chinmaygarde** [comment]
`media_service.dart` has also now been patched to remove the use of `http.readDataPipe`. Another look please.

File: packages/flutter/lib/src/http/http.dart [hunk 2]
```diff
@@ -2,42 +2,49 @@
...
-/// the same server, you should use a single [MojoClient] for all of those requests,
-/// so that the same underlying TCP connection can be re-used (via HTTP pipelining).
-Future<Response> head(dynamic url) {
-  return _withClient/*<Response>*/((MojoClient client) => client.head(url));
-}
+/// the same server, you should use a single [Client] for all of those requests.
+///
+/// For more fine-grained control over the request, use [Request] instead.
+Future<Response> head(dynamic url, {Map<String, String> headers}) =>
+  _withClient((Client client) => client.head(url, headers: headers));
```
Comment (abarth) [inline packages/flutter/lib/src/http/http.dart]: I'm surprised you don't need these generic function annotations.

File: packages/flutter/lib/src/services/asset_bundle.dart [hunk 3]
```diff
@@ -237,9 +176,8 @@ AssetBundle _initRootBundle() {
           context: 'while initialising the root bundle'
         ));
       }
-      return true;
+      return false;
```
Comment (abarth) [inline packages/flutter/lib/src/services/asset_bundle.dart]: This doesn't cause all the tests to fail?

**Review by abarth — APPROVED** [review]
LGTM but give @Hixie a chance to comment.

File: packages/flutter/lib/src/services/asset_bundle.dart (hunk 3 above)
Comment (chinmaygarde) [inline packages/flutter/lib/src/services/asset_bundle.dart]: So the NetworkAssetBundle was being used after all. I brought it back.

Comment (abarth) [inline packages/flutter/lib/src/services/asset_bundle.dart]: I think we want an empty bundle for testing (or probably one that is configurable by the test---but we can start with an empty one).

File: packages/flutter/lib/src/services/image_provider.dart [hunk 4]
```diff
@@ -296,15 +297,36 @@ class NetworkImage extends DataPipeImageProvider<NetworkImage> {
...
-  @override
-  double getScale(NetworkImage key) {
+  Future<ImageInfo> _loadAsync(NetworkImage key) async {
     assert(key == this);
-    return key.scale;
+
+    final Uri resolved = Uri.base.resolve(key.url);
+    final http.Response response = await http.get(resolved);
+    if (response == null || response.statusCode != 200)
+      throw 'Unable to read image data from url.';
```
Comment (Hixie) [inline packages/flutter/lib/src/services/image_provider.dart]: can you throw FlutterError objects instead? We're planning on turning on the lint that whines about non-Error exception objects (like strings).

Comment (Hixie) [inline packages/flutter/lib/src/services/image_provider.dart]: See https://docs.flutter.io/flutter/foundation/FlutterError/FlutterError.html

Comment (Hixie) [inline packages/flutter/lib/src/services/image_provider.dart]: and https://docs.flutter.io/flutter/foundation/FlutterError/message.html

**Comment by Hixie** [comment]
Am I correct in my understanding that the plan is to eventually nuke the entire lib/src/http directory?

If so, please put something to that effect in lib/src/http/README.md and lib/http.dart, and LGTM.

**Comment by chinmaygarde** [comment]
@Hixie I have mocked the http client in the tests as well. Can you take a look at https://github.com/flutter/flutter/pull/5940/commits/793c73ea092ca0bf5c0ef4cc4d3c7d6ed383af05 specifically.

File: packages/flutter/lib/http.dart [hunk 5]
```diff
@@ -3,16 +3,18 @@
...
 /// To use, import `package:flutter/http.dart`.
 ///
-/// This library is based on Dart's `http` package, but differs in that it is a
-/// `mojo`-based HTTP client and does not have a dependency on mirrors.
+/// This library is based on Dart's `http` package, but differs in that it does
+/// not have a dependency on mirrors.
 ///
-/// This library depends only on core Dart libraries as well as the `mojo`,
-/// `mojo_services`, and `sky_services` packages.
+/// TODO(chinmaygarde): The contents of `lib/src/http` will become redundant
```
Comment (Hixie) [inline packages/flutter/lib/http.dart]: make this // rather than /// so it doesn't appear in the public docs

**Comment by Hixie** [comment]
if you wrote a README, you forgot to add it.

LGTM.

File: packages/flutter/lib/http.dart (hunk 5 above)
Comment (chinmaygarde) [inline packages/flutter/lib/http.dart]: Done.

**Comment by chinmaygarde** [comment]
Added the README. Rebased. Landing.

**Comment by chinmaygarde** [comment]
WTH. I only added a README and Travis is failing.

**Pull request merged**